### DISCUSSION TITLE
esp32 wifi range

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12125] [Using MicroPython]
Looking at some of the wifi libs the arduino crowd use I'm wondering if there are any wifi range increasing options exposed in upython? I'm guessing probably not given that espressifs 802.11 LR never seems to have realised much in the way of extra range even with 125kb/s & supposed -105 dbm sensitivity.

I'd like to ditch the 5dBm external antennas in favour of the small pcb antenna on most esp32 boards but I can't see a way to make good the 5 dB difference. I'm finding the wifi gets ratty on most boards once yo get past -70 dBm on the received signal.

**Comment by davefes** [comment]
Using WiFi on a ESP32-CAM I determined that I could send pictures down to about -95dBm.  Have you got a real busy WiFi environment? 
```py
    rssi = sta_if.status('rssi')
    print (f'RSSI =  {rssi} dBm')
    time.sleep_ms(10)

    if ((rssi < -95) or (rssi == 0)):
        print (f'signal level is below -95dBm')
        time.sleep_ms(10)

        sta.active(False)
        sta.disconnect() # or you could get errors
        print (f'machine reset')
        time.sleep_ms(10)

        machine.reset() #  start from scratch
```

**Reply by kjm1102** [reply]
Not really, maybe half a dozen devices on the router wifi. Maybe I should check some other esp32 boards, I took those readings a year or so ago & I've since caught other devices (most notably LTE modems) reporting stale RSSIs. Maybe wifi does the same thing, reporting the RSSI from the initial connect & not actually updating it for sta.status('rssi') until there is a new connection?

 Do you know if we can access the WIFI.PHY.RATE from upython? Seems like the arduino programmers can drop down to 802.11B 1Mbps when they want more range?

**Reply by davefes** [reply]
I don't, maybe @glenn20 knows.

**Reply by glenn20** [reply]
The espnow module provides the `config(rate=XX)` ([docs](https://docs.micropython.org/en/latest/library/espnow.html#espnow.ESPNow.config)) option to set the transmit rate for sending ESP-NOW messages, where XX must be set to a number from the allowed numeric values in [enum wifi_phy_rate_t](https://docs.espressif.com/projects/esp-idf/en/v4.4.1/esp32/api-reference/network/esp_wifi.html#_CPPv415wifi_phy_rate_t). This uses the [esp_wifi_config_espnow_rate()](https://docs.espressif.com/projects/esp-idf/en/v4.4.1/esp32/api-reference/network/esp_wifi.html#_CPPv427esp_wifi_config_espnow_rate16wifi_interface_t15wifi_phy_rate_t) IDF function call.

The  IDF also includes the [esp_wifi_config_80211_tx_rate()](https://docs.espressif.com/projects/esp-idf/en/v4.4.1/esp32/api-reference/network/esp_wifi.html#_CPPv429esp_wifi_config_80211_tx_rate16wifi_interface_t15wifi_phy_rate_t) function to set the 802.11 tx rate, but access to this is not provided by micropython.

**Reply by glenn20** [reply]
However, the LR_MODE (`sta.config(protocol=network.MODE_LR)`) is supported (the [esp32 LR docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi.html#long-range-lr) suggest this gains about a 2 times increase in range). However, I have no experience with using it. I have seen reports from users that have, but none that actually measure the rate of dropped packets at increasing range.

**Comment by kjm1102** [comment]
Is it me or is that doc contradictory? In the LR Compatibility section it
says LR is incompatible with 802.11 because the beacon is sent in LR mode
but then in the LR impacts it says the opposite?

**Reply by glenn20** [reply]
> For LR-enabled AP of ESP32, it is incompatible with traditional 802.11 mode, because the beacon is sent in LR mode.
- This means that if the AP is in LR mode, it is not compatible with a non-LR client (ie. only ESP32 devices configured to support LR can connect to an AP operating in LR mode.

I **assume** the comments in the "LR Impacts" section are referring to clients operating with LR mode supported (ie. the STA interface).

**Comment by bidrohini** [comment]
I am not getting any suggestion other than adding an external antenna: https://www.youtube.com/watch?v=hmPjy6LZQFo

**Comment by peterhinch** [comment]
You're up against basic physics here. The transmitter power, the receiver sensitivity, and the modulation scheme are fixed. The only things you can change are the antennae at either end of the link. Replacing a chip antenna with something more efficient such as a 1/4 wave duck may help - but efficiency in terms of converting electrical power to RF can't get better than 100%. Beyond that the only option is to use directional antennae which direct the available RF power into smaller regions. In principle this can yield big improvements but for obvious reasons a directional aerial at both ends is only really suitable for fixed links.

**Reply by davefes** [reply]
But if the modulation scheme can be changed to a lower datatrate along with suitable processing in the receiver (to narrow the bandwidth) then more range should be expected. 

In ESPNow does `config(rate=XX)` actually change anything in the receiver, as well?  Searching the 726 page tech reference for keywords like sensitivity, selectivity, BER, even RF does not turn up anything relevant.

Does anyone know what receiver architecture is used in these things?

**Comment by kjm1102** [comment]
Using an outdoor AP with 8db antenna & esp32 with 5db whip longest distance I can send a photo reliably is ~700m (esp32 reports ~-90dBm). I assume wifi has the smarts to try different data rates but the only metric I can find exposed in upython is the rssi. Sure would be nice to force 1Mbps with regular wifi like the LR/Now modes just to see the range difference to a regular outdoor wifi AP. Wifi Halow is still as dear as poison.
